### PR TITLE
sanitizer issues

### PR DESCRIPTION
**Pull request by aautushka**
Fix a bunch of errors reported by code sanitizers

memcmp: passing nullptr, when non-null expected
```
/tests/crypto/base_n_tests.cpp:56:5: runtime error: null pointer passed as argument 1, which is declared to never be null
/src/crypto/base36.cpp:79:59: runtime error: null pointer passed as argument 2, which is declared to never be null
/tests/crypto/base_n_tests.cpp:79:5: runtime error: null pointer passed as argument 1, which is declared to never be null
/src/crypto/base58.cpp:635:45: runtime error: null pointer passed as argument 2, which is declared to never be null
/tests/crypto/base_n_tests.cpp:84:5: runtime error: null pointer passed as argument 1, which is declared to never be null
```
'blowfish_chain_test': using memory after allocated heap
```
=================================================================
==15254==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60300000ebdd at pc 0x000000eee169 bp 0x7ffc2d831220 sp 0x7ffc2d831210
READ of size 1 at 0x60300000ebdd thread T0
    #0 fc::BytesToBlock(unsigned char const*, fc::sblock&)
    #1 fc::blowfish::encrypt(unsigned char const*, unsigned char*, unsigned long, int)
    #2 fc_crypto::blowfish_chain_test::test_method()
    #3 fc_crypto::blowfish_chain_test_invoker()

```
'websocket_test': uncatched exceptions
```
unknown location(0): fatal error in "websocket_test": unknown type
/tests/network/http/websocket_test.cpp(52): last checkpoint
```
'fc_crypto*' test_stream<>: uninitialized variable (hash)
```
==17672== Conditional jump or move depends on uninitialised value(s)
==17672==    at 0x4C33CF2: __memcmp_sse4_1 (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==17672==    by 0xB63764: fc::operator==(fc::ripemd160 const&, fc::ripemd160 const&) (ripemd160.cpp:98)
==17672==    by 0xA422C9: void test_stream<fc::ripemd160>() (sha_tests.cpp:84)
==17672==    by 0xA3BC6E: fc_crypto::ripemd160_test::test_method() (sha_tests.cpp:108)
==17672==    by 0xA3BAC3: fc_crypto::ripemd160_test_invoker() (sha_tests.cpp:99)
```

### CODE REVIEW

File: include/fc/static_variant.hpp [hunk 1]
```diff
@@ -181,7 +181,7 @@ class static_variant {
     static_assert(impl::type_info<Types...>::no_reference_types, "Reference types are not permitted in static_variant.");
     static_assert(impl::type_info<Types...>::no_duplicates, "static_variant type arguments contain duplicate types.");
 
-    int _tag;
+    size_t _tag;
```
Comment (abitmore) [inline include/fc/static_variant.hpp]: How does `size_t` get serialized? It's not safe to change the type from signed to unsigned when it's already in production, since it may break consensus. Safer to change to `int64_t` as already done in Steem: https://github.com/steemit/fc/commit/b4fd91d889796c6aee84428427552046e5316392. Related discussion in https://github.com/steemit/fc/pull/37.

File: src/rpc/cli.cpp [hunk 2]
```diff
@@ -160,20 +160,20 @@ char* my_generator(const char* text, int state)
 }
 
 
+#ifdef HAVE_READLINE
```
Comment (abitmore) [inline src/rpc/cli.cpp]: This changes seems not correct, as it removed definition/implementation of a function, need to make sure that it's not used somewhere else.

File: src/thread/thread.cpp [hunk 3]
```diff
@@ -113,8 +113,10 @@ namespace fc {
       if( my )
       {
         // wlog( "calling quit() on ${n}",("n",my->name) );
-        quit(); // deletes `my`
+        quit();
       }
+
+      delete my;
```
Comment (abitmore) [inline src/thread/thread.cpp:119]: If can delete `my` anyway, why need to check `if( my )` above?

File: tests/crypto/base_n_tests.cpp [hunk 4]
```diff
@@ -76,17 +76,14 @@ static void test_58( const std::string& test, const std::string& expected )
...
     size_t len = fc::from_base58( enc1, buffer, 64 );
     BOOST_CHECK( len <= 64 );
-    BOOST_CHECK( !memcmp( vec.data(), buffer, len ) );
+    BOOST_CHECK( vec.empty() || !memcmp( vec.data(), buffer, len ) );
     if ( len > 10 ) {
-        try {
-            len = fc::from_base58( enc1, buffer, 10 );
-            BOOST_CHECK( len <= 10 );
-        } catch ( fc::exception expected ) {}
+        BOOST_CHECK_THROW(fc::from_base58( enc1, buffer, 10 ), fc::exception);
```
Comment (abitmore) [inline tests/crypto/base_n_tests.cpp:86]: Why this change? It's totally different logic.

File: tests/crypto/sha_tests.cpp [hunk 5]
```diff
@@ -75,7 +75,7 @@ void test_big( const std::string& expected ) {
 
 template<typename H>
 void test_stream( ) {
-    H hash( TEST1 );
+    H hash = H::hash( TEST1 );
```
Comment (abitmore) [inline tests/crypto/sha_tests.cpp:78]: This is different.

File: tests/network/http/websocket_test.cpp [hunk 6]
```diff
@@ -39,7 +39,7 @@ BOOST_AUTO_TEST_CASE(websocket_test)
         try {
             c_conn->send_message( "again" );
             BOOST_FAIL("expected assertion failure");
-        } catch (const fc::assert_exception& e) {
+        } catch (const fc::exception& e) {
```
Comment (abitmore) [inline tests/network/http/websocket_test.cpp:42]: Why this?

**Comment by abitmore** [comment]
I'd recommend that do one related changes in one commit, but not put all them together.

**Review by ryanRfox — COMMENTED** [review]
I have reviewed the code changes. However, will await further comments from @aautushka in response to those offered by @abitmore inline.

File: src/rpc/cli.cpp (hunk 2 above)
Comment (aautushka) [reply src/rpc/cli.cpp]: This fixes the unused function warning when compiled without readline. There is but a single call of the function protected by another HAVE_READLINE. See cli::getline@cli.cpp

File: tests/crypto/base_n_tests.cpp (hunk 4 above)
Comment (aautushka) [reply tests/crypto/base_n_tests.cpp:86]: Actually the test expects fc::from_base58 to throw. We're just using the standard Boost way rather than a quite obscure try{ BOOST_CHECK(); } catch(){}

In essence  it's exactly the same test.

File: tests/crypto/sha_tests.cpp (hunk 5 above)
Comment (aautushka) [reply tests/crypto/sha_tests.cpp:78]: Yes, it is. We believe there is a bug in sha_tests. You should look at the hash class constructor (ripemd160 for example) which accepts the hex-encoded hash string. TEST1 on the other hand is declared this way:
```
static const std::string TEST1("abc")
```
which is apparently not a ripemd160 hash (too short a string for ripemd160), as a result it leads to uninitialized memory usage. 

Anyways, the test itself is all about object serialization/deserialization, and it should not affect its success if the hash object is correctly constructed indeed or is in some undefined state (due to inititialized memory).

File: tests/network/http/websocket_test.cpp (hunk 6 above)
Comment (aautushka) [reply tests/network/http/websocket_test.cpp:42]: We've just fixed a failing test. It seems that at some point the underlying logic has changed to no longer throw a fc::assert_exception and the test went dead. 

Please note the pattern: the code is not using BOOST_EXPECT_THROW as it could have thus harming readability.

File: src/thread/thread.cpp (hunk 3 above)
Comment (aautushka) [reply src/thread/thread.cpp:119]: Sorry, I don't frankly get your point. It's perfectly safe in C++ to delete nullptr - it would do absolutely nothing. 

We absolutely need to check the my for null before calling quit() since the pointer would be dereferenced inside quit() thus causing a SIGSEGV.

**Review by aautushka — COMMENTED** [review]
@abitmore appreciate your comments. Next time we'll do our best to submit smaller pull requests.

File: include/fc/static_variant.hpp (hunk 1 above)
Comment (aautushka) [reply include/fc/static_variant.hpp]: Pretty sure it's as safe as it could get. The tag in static_variant gets assigned at compile time, it is protected with compile time asserts. The tag should not go beyond the [0, N) boundary and if it goes  somehow (presumably due to broken deserialization or some intentional abuse) you'll get a runtime exception. 

But perhaps you have better perspective and we'll surely do as you suggest.

File: src/rpc/cli.cpp (hunk 2 above)
Comment (abitmore) [reply src/rpc/cli.cpp]: Good to know. Thanks.

File: src/thread/thread.cpp (hunk 3 above)
Comment (abitmore) [reply src/thread/thread.cpp:119]: Oh, I didn't know `nullptr` can be safely deleted in C++ 11. My knowledge about `delete` is still on older version. Thanks.

File: tests/crypto/base_n_tests.cpp (hunk 4 above)
Comment (abitmore) [reply tests/crypto/base_n_tests.cpp:86]: The old test code will still pass if `fc::from_base58` doesn't throw. In this case will check `len <= 10`. So this is different test. I don't know whether the old test code is buggy though.

File: tests/network/http/websocket_test.cpp (hunk 6 above)
Comment (abitmore) [reply tests/network/http/websocket_test.cpp:42]: Makes sense. Bug fixes are OK.
However, you named this PR as fixes to "sanitizer issues" so I mostly checked if your code has changed logic.

File: include/fc/static_variant.hpp [hunk 7]
```diff
@@ -326,15 +327,14 @@ class static_variant {
     }
 
     static int count() { return impl::type_info<Types...>::count; }
-    void set_which( int w ) {
-      FC_ASSERT( w >= 0 );
+    void set_which( tag_type w ) {
```
Comment (abitmore) [inline include/fc/static_variant.hpp]: Shouldn't remove `FC_ASSERT( w >= 0 );` if the type is changed to `int64_t`.

**Comment by abitmore** [comment]
@ryanRfox I do appreciate your comment, but I expected that you'll find more qualified developers to review this PR and hopefully find more issues that I didn't find, but not just a comment on my comments. Thanks.

**Comment by abitmore** [comment]
> @abitmore appreciate your comments. Next time we'll do our best to submit smaller pull requests.

I mean smaller commits. Grouping them in one PR is fine. Thanks.

File: include/fc/static_variant.hpp (hunk 1 above)
Comment (abitmore) [reply include/fc/static_variant.hpp]: Yes, maybe there are intentional abuses, e.g. comparison, `0<-1` is `false` for `int64_t` but `true` for `uint64_t` (perhaps runtime error for `size_t`?)

File: src/thread/mutex.cpp [hunk 8]
```diff
@@ -15,7 +15,7 @@ namespace fc {
   mutex::~mutex() {
     if( m_blist ) 
     {
-      context* c = m_blist;
```
Comment (abitmore) [inline src/thread/mutex.cpp:18]: I'd recommend that move it to inside `#if 0`, although I don't know why there is a `#if 0`.

Comment (aautushka) [reply src/thread/mutex.cpp:18]: Must be some long forgotten debug stuff

Comment (aautushka) [reply src/thread/mutex.cpp:18]: bitshares-fc definitely needs to be cleaned up

File: tests/network/http/websocket_test.cpp (hunk 6 above)
Comment (aautushka) [reply tests/network/http/websocket_test.cpp:42]: Yeah, we lumped in lots of warning fixes as well... We are mostly using GCC7 and it's being very loud about every minute issue it thinks it has with the code.

File: tests/crypto/base_n_tests.cpp (hunk 4 above)
Comment (aautushka) [reply tests/crypto/base_n_tests.cpp:86]: Hmm... You see, the test either throws or it does not, it could not do both, and if it does both, then the test is indeterministic and must be fixed either way. 

Here is the very exception being tested. The function most definitely throws and never reaches BOOST_CHECK()
```
(gdb) bt
#0  0x00007ffff69faf1d in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x0000000000a91143 in fc::from_base58 (base58_str="2zuFXTJSTRK6ESktqhM2QDBkCnH1U46CnxaD", out_data=0x7fffffffbc40 "ABCDEF
GHIJKLMNOPQRSTUVWXYZ\375\367\377\177", out_data_len=10) at /home/aautushka/projects/bitshares-fc/src/crypto/base58.cpp:634
#2  0x0000000000962144 in fc_crypto::test_58 (test="ABCDEFGHIJKLMNOPQRSTUVWXYZ", expected="2zuFXTJSTRK6ESktqhM2QDBkCnH1U46Cnx
aD") at /home/aautushka/projects/bitshares-fc/tests/crypto/base_n_tests.cpp:87
#3  0x00000000009626f1 in fc_crypto::base58_test::test_method (this=0x7fffffffc10f) at /home/aautushka/projects/bitshares-fc/
tests/crypto/base_n_tests.cpp:96
#4  0x00000000009624b1 in fc_crypto::base58_test_invoker () at /home/aautushka/projects/bitshares-fc/tests/crypto/base_n_test
s.cpp:92
#5  0x000000000095e301 in boost::detail::function::void_function_invoker0<void (*)(), void>::invoke (function_ptr=...) at /op
t/boost_1_63_0/include/boost/function/function_template.hpp:118
#6  0x0000000000c60eeb in boost::detail::function::function_obj_invoker0<boost::detail::forward, int>::invoke(boost::detail::
function::function_buffer&) ()
#7  0x0000000000c6057d in boost::execution_monitor::catch_signals(boost::function<int ()> const&) ()
#8  0x0000000000c60656 in boost::execution_monitor::execute(boost::function<int ()> const&) ()
#9  0x0000000000c60daa in boost::execution_monitor::vexecute(boost::function<void ()> const&) ()
#10 0x0000000000c2dd9b in boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsig
ned int) ()
#11 0x0000000000c0bd66 in boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned int, boost::unit_test
::framework::state::random_generator_helper const*) ()
#12 0x0000000000c0c042 in boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned int, boost::unit_test
::framework::state::random_generator_helper const*) ()
#13 0x0000000000c0c042 in boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned int, boost::unit_test
::framework::state::random_generator_helper const*) ()
#14 0x0000000000c058d2 in boost::unit_test::framework::run(unsigned long, bool) ()
#15 0x0000000000c2bf7f in boost::unit_test::unit_test_main(boost::unit_test::test_suite* (*)(int, char**), int, char**) ()
#16 0x00007ffff60a1830 in __libc_start_main (main=0x95a790 <main>, argc=3, argv=0x7fffffffd968, init=<optimized out>, fini=<o
ptimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd958) at ../csu/libc-start.c:291
#17 0x000000000095b469 in _start ()
(gdb)
```

**Review by abitmore — DISMISSED** [review]
Hope others will review.

**Comment by xeroc** [comment]
Not being a C++ developer i still took a chance to read through this patch and couldn't find anything strange with it. Thanks for the good in-line discussion, I learned something from it!

File: src/crypto/aes.cpp [hunk 9]
```diff
@@ -30,6 +30,7 @@ struct aes_encoder::impl
 aes_encoder::aes_encoder()
 {
   static int init = init_openssl();
+  (void)init;
```
Comment (pmconrad) [inline src/crypto/aes.cpp:33]: Is this to suppress an "unused variable" warning?

File: src/thread/mutex.cpp (hunk 8 above)
Comment (pmconrad) [reply src/thread/mutex.cpp:18]: I'd remove the whole #if 0 block. Doesn't improve readability.

File: src/crypto/aes.cpp (hunk 9 above)
Comment (xeroc) [reply src/crypto/aes.cpp:33]: I guess so. Have seen that often in pure C projects that have tight compile settings.

**Review by pmconrad — APPROVED** [review]
